use os.user instead of reading /etc/group

kaworu opened the pull request:
Use `os/user` from Go stdlib to map the group name to the gid instead of searching through /etc/group. Couple of comments changes on the way.

kaworu committed:
use os.user instead of reading /etc/group

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

tklauser commented on pkg/api/socket.go:
```diff
@@ -32,30 +30,15 @@ var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "api")
 
 // GetGroupIDByName returns the group ID for the given grpName.
 func GetGroupIDByName(grpName string) (int, error) {
```
While at it, I think this func can be unexported since its only user is in the same package (`api.SetDefaultPermissions`)

sayboras replied:
just curious, based on https://github.com/golang/go/blob/78c20c81aaa7d8d8a4aa2132806f51fb68f33a1e/src/os/user/cgo_lookup_unix.go#L6, there is another tag for cgo, as we disabled cgo in our build, do we need to add _osusergo_ tag ?

tklauser replied:
@sayboras We e.g. build the `proxylib` with cgo but IMO we should make sure to still use the Go-only implementation even when building with cgo, so the behavior is consistent. Thus I think it's safer to always explicity specify `osusergo`.

kaworu replied:
My understanding is that the osusergo tag effect is to use the go implementation when cgo is enabled, and has no effect when cgo is disabled.

_cgo_lookup_unix.go_ has `// +build cgo,!osusergo` meaning `cgo && !osusergo`  
_lookup_unix.go_ has `// +build !cgo osusergo` meaning `!cgo || osusergo`

sayboras replied:
@kAworu @tklauser you are right. I overlooked this line https://github.com/golang/go/blob/master/src/os/user/lookup_unix.go#L6 --> it's OR condition :tada:

tklauser reviewed (changes requested):
According to the package godoc for `os/user` (https://github.com/golang/go/blob/78c20c81aaa7d8d8a4aa2132806f51fb68f33a1e/src/os/user/user.go#L13-L15), we should probably use the `osusergo` build tag to make sure we're not using the cgo-based implementation under any circumstances. In general we want to avoid cgo in cilium.

aanm commented on pkg/api/socket.go:
```diff
@@ -30,34 +28,19 @@ import (
-	f, err := os.Open(GroupFilePath)
+// getGroupIDByName returns the group ID for the given grpName.
+func getGroupIDByName(grpName string) (int, error) {
+	group, err := user.LookupGroup(grpName)
```
This is ridiculously simple compared to the older code :smile:

tklauser commented on Makefile.defs:
```diff
@@ -132,7 +132,7 @@ ifneq ($(LOCKDEBUG),)
 endif
 
-GO_BUILD_FLAGS += -ldflags '$(GO_BUILD_LDFLAGS) $(EXTRA_GO_BUILD_LDFLAGS)' $(EXTRA_GO_BUILD_FLAGS)
+GO_BUILD_FLAGS += -tags osusergo -ldflags '$(GO_BUILD_LDFLAGS) $(EXTRA_GO_BUILD_LDFLAGS)' $(EXTRA_GO_BUILD_FLAGS)
```
Let's add `-tags osusergo` to `$GO_TEST_FLAGS` as well so were guaranteed to be using the non-cgo implementation in the tests as well.

kaworu replied:
Of course, my mistake.

tklauser replied:
No worries. Might make sense to introduce a separate `Makefile` variable for Go tags (e.g. `lockdebug` above) at some point which is then appended to both build and test flags. But that is of course outside the scope of this PR :smiley:

kaworu replied:
Adding to `$GO_TEST_FLAGS` also adds to `$GO_CLEAN` which is probably not an issue but weird looking:

```
go clean -mod=vendor -tags osusergo
```

tklauser replied:
I think this is a bug in the definition of `GO_CLEAN`:

Currently it says:
```make
GO_CLEAN = $(GO) clean $(GO_TEST_FLAGS)
```
but I think it should be:
```make
GO_CLEAN = $(GO) clean $(GO_CLEAN_FLAGS)
```

tklauser reviewed (approved):
Nice :tada:

kaworu committed:
make: uses $GO_CLEAN_FLAGS in $GO_CLEAN

Before this patch, $GO_CLEAN_FLAGS was unused and $GO_CLEAN used
$GO_TEST_FLAGS.

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

kaworu committed:
make: refactor the way "join with comma" is done

Moved $(empty), $(space), and $(comma) from Makefile to Makefile.defs
and added the $(join-with-comma) expression.

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

kaworu committed:
make: refactor the way -tags is given to go build/test

While go build accept several -tags, go test doesn't:

    go test: tags flag may be set only once

This patch introduce $GO_TAGS_FLAGS to aggregate the tags and provide
them as a single comma-separated list to go build/test.

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

errordeveloper commented on Makefile.defs:
```diff
@@ -4,6 +4,12 @@
+empty :=
+space := $(empty) $(empty)
+comma := ,
+join-with-comma = $(subst $(space),$(comma),$(strip $1))
```
can you use `foreach`?

kaworu replied:
Joining using `foreach` imply handling either the first (or last) iteration differently in order to avoid an extra leading (respectively trailing) comma correct?

errordeveloper commented on Makefile.defs:
```diff
@@ -4,6 +4,12 @@
 SHELL := /bin/bash
 .SHELLFLAGS := -eu -o pipefail -c
 
+# Some GNU Make trickery to basically achieve s/ /,/g
```
Why no just say what it does?
```suggestion
# define a function that replaces spaces with comas in a list
```

vadorovsky reviewed (approved):
Looks good. Thanks!

sayboras commented on operator/Makefile:
```diff
@@ -9,14 +9,14 @@ TARGETS := cilium-operator cilium-operator-generic cilium-operator-aws cilium-op
-cilium-operator-generic: OPERATOR_TAGS=ipam_provider_operator
-cilium-operator-aws: OPERATOR_TAGS=ipam_provider_aws
-cilium-operator-azure: OPERATOR_TAGS=ipam_provider_azure
+cilium-operator: GO_TAGS_FLAGS+=ipam_provider_aws,ipam_provider_azure,ipam_provider_operator
```
:exploding_head: :100: 

I didn't know that this is possible.

kaworu replied:
I believe it is specific to GNU Make

tklauser replied:
The magic of target-specific variable values in GNU make :sparkles: :smile: https://www.gnu.org/software/make/manual/html_node/Target_002dspecific.html

sayboras replied:
thanks both for info, I can simplify some of the stuffs I have at work with this :100:
